sldns/str2wire.c

```diff
@@ -615,6 +614,113 @@ sldns_affix_token(sldns_buffer* strbuf, char* token, size_t* token_len,
 	return 1;
 }
 
+static int sldns_str2wire_svcparam_key_cmp(const void *a, const void *b)
+{
+	return sldns_read_uint16(*(uint8_t**) a)
+	     - sldns_read_uint16(*(uint8_t**) b);
+}
+
+/**
+ * Add constraints to the SVCB RRs which involve the whole set
+ */
+static int sldns_str2wire_check_svcbparams(uint8_t* rdata, uint16_t rdata_len)
+{
+	size_t   nparams = 0, i, j;
+	uint8_t  new_rdata[65536];
+	uint8_t* new_rdata_ptr = new_rdata;
+	uint8_t* svcparams[64];
+	uint8_t* mandatory = NULL;
+	uint8_t* rdata_ptr = rdata;
+	uint16_t rdata_remaining = rdata_len;
+
+	/* find the SvcParams */
+	while (rdata_remaining) {
+		uint16_t svcbparam_len;
+
+		svcparams[nparams] = rdata_ptr;
+		if (rdata_remaining < 4)
+			return LDNS_WIREPARSE_ERR_GENERAL;
+		svcbparam_len = sldns_read_uint16(rdata_ptr + 2);
+		rdata_remaining -= 4;
+		rdata_ptr += 4;
+
+		if (rdata_remaining < svcbparam_len)
+			return LDNS_WIREPARSE_ERR_GENERAL;
+		rdata_remaining -= svcbparam_len;
+		rdata_ptr += svcbparam_len;
+
+		nparams += 1;
+		if (nparams > sizeof(svcparams)/8)
+			return LDNS_WIREPARSE_ERR_SVCB_TOO_MANY_PARAMS;
+	}
+
+	/* In draft-ietf-dnsop-svcb-https-05 Section 7:
+	 *
+	 *     In wire format, the keys are represented by their numeric
+	 *     values in network byte order, concatenated in ascending order.
+	 */
+	qsort((void *)svcparams
+	     ,nparams
+	     ,sizeof(uint8_t*)
+	     ,sldns_str2wire_svcparam_key_cmp);
+
+	/* In draft-ietf-dnsop-svcb-https-05 Section 7:
+	 *
+	 *     Keys (...) MUST NOT appear more than once.
+	 *
+	 * If they key has already been seen, we have a duplicate
+	 */
+	for (i = 0; i < nparams; i++) {
+		uint16_t key = sldns_read_uint16(svcparams[i]);
+
+		if (i + 1 < nparams && key == sldns_read_uint16(svcparams[i+1]))
+			return LDNS_WIREPARSE_ERR_SVCB_DUPLICATE_KEYS;
+
+		if (key == SVCB_KEY_MANDATORY)
+			mandatory = svcparams[i];
+	}
+
+	/* 4. verify that all the SvcParamKeys in mandatory are present */
+	if (mandatory) {
+
+		/* Divide by sizeof(uint16_t)*/
+		uint16_t mandatory_len = sldns_read_uint16(mandatory + 2) >> 1;
```

Comment:
Is using `>> 1` the most readable way to write divide by 2?
```suggestion
		uint16_t mandatory_len = sldns_read_uint16(mandatory + 2) / sizeof(uint16_t);
```
Also perhaps rename `mandatory_len` into `mandatory_nkeys`?